src/HttpUpgrade.h

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright (C) 1996-2019 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#ifndef SQUID_HTTP_UPGRADE_H
+#define SQUID_HTTP_UPGRADE_H
+
+#include "acl/forward.h"
+#include "base/StringView.h"
+#include "sbuf/forward.h"
+
+#include <map>
+
+/// a reference to a protocol name[/version] string; no 0-termination is assumed
+class ProtocolView
+{
+public:
+    ProtocolView(const char * const start, const size_t len);
+    explicit ProtocolView(const StringView &proto);
+    explicit ProtocolView(const SBuf &proto);
+
+    StringView name; ///< everything up to (but excluding) the first slash('/')
+    StringView version; ///< everything after the name, including the slash('/')
+};
+
+std::ostream &operator <<(std::ostream &, const ProtocolView &);
+
+// HTTP is not explicit about case sensitivity of Upgrade protocol strings, but
+// there are bug reports showing different case variants used for WebSocket. We
+// conservatively preserve the received case and compare case-sensitively.
+
+/// Either b has no version restrictions or both have the same version.
+/// For example, "ws/1" is in "ws" but "ws" is not in "ws/1".
+inline bool
+vAinB(const ProtocolView &a, const ProtocolView &b)
+{
+    // Optimization: Do not assert(a.name == b.name).
+    return b.version.empty() || (a.version == b.version);
+}
+
+/// Allows or blocks HTTP Upgrade protocols (see http_upgrade_request_protocols)
+class HttpUpgradeProtocolAccess
```

Comment:
The naming of this file vs the primary class name violates our coding guidelines.
This should all be in src/HttpUpgradeProtocolAccess.h for now.

Reply:
Commit b96234394d renames HttpUpgrade.[cc,h] files to HttpUpgradeProtocolAccess.[cc,h].

I also started moving the ProtocolView class inside HttpUpgradeProtocolAccess class to be 
real sub-class, but finally I decided to not commit this patch.
If required I can do it later..

Reply:
This is fine. Thank you.